### PR TITLE
remove RX as the primary pipe for gRPC communications

### LINKED ISSUE #8280: Proposal: remove RX from the primary gRPC message pipe
#### What problem would the feature you're requesting solve? Please describe.

Profiling demonstrated a performance bottleneck associated with the way that RX is used as part of the gRPC messaging pipe. At the moment there is a single common RX hub, and all events to all workers go through this hub, using `Where` filters at the observer to filter data back out to the correct worker. In addition to the significant overheads, this limits the ability to use parallel connections to workers (to reduce head-of-line blocking and general comms congestion), because in the `Where(...)` model, all connections to the same worker would forward the same message, causing multiple delivery.

#### Describe the solution you'd like

Instead of a single centralised hub, use bespoke directed channels - presumably `Channel<T>` - as the metaphor to convey messages to individual workers.

## Describe alternatives you've considered

n/a

#### Additional context

n/a

### PR DESCRIPTION
**Pull request by mgravell**
### Issue describing the changes in this PR

resolves #8280; performance locally is a 5x+ speedup in throughput

the main crux of the new flow is described in src/WebJobs.Script.Grpc/Eventing/GrpcEventExtensions.cs, but in short: we establish a per-worker `Channel<InboundGrpcEvent>` and `Channel<OutboundGrpcEvent>` to handle the backlog; instead of publishing to RX, we publish to the writer of the relevant queue. Separately, we have an async worker deque data from the queues, and process accordingly. This is much more direct, avoids a *lot* of RX machinery, and creates isolation between workers.

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

**Comment by fabiocav** [comment]
@mgravell apologies for the delay on this. We'll get eyes on this ASAP. Would it be possible to have this rebased to deal with the conflicts? 

Thanks!

**Comment by mgravell** [comment]
Will do

**Comment by pragnagopa** [comment]
Tagging @safihamid @paulbatum  - fyi 

@fabiocav  / @brettsam  - would be good to get test E2E on private stamp before merging this PR

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs [hunk 1]
```diff
@@ -47,6 +48,7 @@ public IRpcWorkerChannel Create(string scriptRootPath, string runtime, IMetricsL
                 throw new InvalidOperationException($"WorkerCofig for runtime: {runtime} not found");
             }
             string workerId = Guid.NewGuid().ToString();
+            _eventManager.AddGrpcChannels(workerId); // prepare the inbound/outbound dedicated channels
```
Comment (safihamid) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:51]: @pragnagopa can you confirm this is being called during placeholder mode and shouldn't be impacting cold start?

Also how is this going to impact isolated cold start when there are no placeholders?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 2]
```diff
@@ -1015,5 +1165,75 @@ private void AddSample<T>(List<T> samples, T sample)
...
+                }
+            }
+
+            private void OnTimeout()
+            {
+                if (MakeComplete() && _faultHandler != null)
+                {
+                    try
+                    {
+                        throw new TimeoutException();
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:1310]: Do we need to throw an new exception just to catch it and handle? Could we just do `_faultHandler(timeout);` in the first try?

**Comment by mgravell** [comment]
@fabiocav  I *believe* that's merged; it looks like some cheese moved around the topic of "function load collections", which made the tests merge a bit messy, but: should work

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 3]
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
 
         internal RpcWorkerConfig Config => _workerConfig;
 
+        private void ProcessItem(InboundGrpcEvent msg)
+        {
+            // note this is a QUWI entry-point
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: We should spell out `QueueUserWorkItem` here -- I had to look it up :-)

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 4]
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
+                }
+                while (await _inbound.WaitToReadAsync())
+                {
+                    while (_inbound.TryRead(out var msg))
+                    {
+                        if (debug)
+                        {
+                            _workerChannelLogger.LogDebug("[channel] received {0}: {1}", msg.WorkerId, msg.MessageType);
+                        }
+                        ThreadPool.QueueUserWorkItem(_processInbound, msg);
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:242]: Curious: why this over `Task.Run()`?

File: src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs]: Remove?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 6]
```diff
@@ -99,31 +102,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
-                .Where(msg => msg.WorkerId == _workerId);
-
-            _eventSubscriptions.Add(_inboundWorkerEvents
-                .Where(msg => msg.IsMessageOfType(MsgType.RpcLog) && !msg.IsLogOfCategory(RpcLogCategory.System))
-                .Subscribe(Log));
-
-            _eventSubscriptions.Add(_inboundWorkerEvents
-                .Where(msg => msg.IsMessageOfType(MsgType.RpcLog) && msg.IsLogOfCategory(RpcLogCategory.System))
-                .Subscribe(SystemLog));
+            if (eventManager.TryGetGrpcChannels(workerId, out var inbound, out var outbound))
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Should we use `_eventManager` here?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 7]
```diff
@@ -99,31 +102,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
+            {
+                _outbound = outbound.Writer;
+                _inbound = inbound.Reader;
+                // note: we don't start the read loop until StartWorkerProcessAsync is called
+            }
 
             _eventSubscriptions.Add(_eventManager.OfType<FileEvent>()
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:122]: Do we also want this to be handled differently? I think it's the only one still using `_eventSubscriptions`

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 8]
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
 
+        private void ProcessItem(InboundGrpcEvent msg)
+        {
+            // note this is a QUWI entry-point
+            try
+            {
+                switch (msg.MessageType)
+                {
+                    case MsgType.RpcLog:
+                        switch (msg.Message.RpcLog.LogCategory)
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: nit: might be cleaner to use the new switch expression: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 9]
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
+            lock (queue)
+            {
+                // while we have the lock, discard any dead items (to prevent unbounded growth on stall)
+                while (queue.TryPeek(out var next) && next.IsComplete)
+                {
+                    queue.Dequeue();
+                }
+                for (int i = 0; i < count; i++)
+                {
+                    var newItem = i == count - 1 && timeout != TimeSpan.Zero
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: nit: for readability
```suggestion
                    var newItem = (i == count - 1) && (timeout != TimeSpan.Zero)
```

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 10]
```diff
@@ -827,9 +937,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
...
+                var pending = _outbound.WriteAsync(evt);
+                if (pending.IsCompleted)
+                {
+                    try
+                    {
+                        pending.GetAwaiter().GetResult(); // ensure observed, for IVTS reasons
+                    }
+                    catch
+                    {
+                        // suppress failure
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:1007]: Just wondering why we want to suppress the failure here

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 11]
```diff
@@ -1002,5 +1158,75 @@ private void AddSample<T>(List<T> samples, T sample)
                 samples.Add(sample);
             }
         }
+
+        private sealed class PendingItem
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:1256]: Is this essentially a pending grpc message / event we want to handle? I think a more descriptive name could be helpful. If we change the name here it might also be good to make sure other methods such as `ProcessItem` match i.e. ProcessGrpcEvent

Also wondering if we should move this into it's own file

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 12]
```diff
@@ -217,6 +329,11 @@ internal void SendWorkerInitRequest(GrpcEvent startEvent)
                 initRequest.Capabilities.Add(RpcWorkerConstants.FunctionDataCache, "true");
             }
 
+            // advertise that we support multiple streams, and hint at a number; with this flag, we allow
+            // clients to connect multiple back-hauls *with the same workerid*, and rely on the internal
+            // plumbing to make sure we don't process everything N times
+            initRequest.Capabilities.Add(RpcWorkerConstants.MultiStream, "10"); // TODO: make this configurable
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:336]: Can we make the number ("10") a readonly constant for now?

Is the TODO for this PR or future work? If future work, can we create a github issue for this and mention it here?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 13]
```diff
@@ -827,9 +937,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
...
+            }
+            static async Task ObserveEventually(ValueTask valueTask)
+            {
+                try
+                {
+                    await valueTask;
+                }
+                catch
+                {
+                    // nowehere to log
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]:
```suggestion
                    // no where to log
```

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs (hunk 1 above)
Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:51]: @pragnagopa poke for update :)

Comment (brettsam) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:51]: @safihamid -- Can you take this and run it in a private stamp to make sure there's no cold start regressions?

File: src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs (hunk 5 above)
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs]: Wondering why this is commented out and if we have plans to use this in this PR?

File: src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs [hunk 14]
```diff
@@ -33,64 +31,48 @@ public FunctionRpcService(IScriptEventManager eventManager, ILogger<FunctionRpcS
...
                     // GRPC does not accept cancellation tokens for individual reads, hence wrapper
                     var requestTask = requestStream.MoveNext(CancellationToken.None);
-                    var completed = await Task.WhenAny(cancelSource.Task, requestTask);
-                    return completed.Result;
-                };
+                    return Task.WhenAny(cancelSource.Task, requestTask);
+                }
 
-                if (await messageAvailable())
+                if (await await MoveNextAsync(requestStream, cancelSource))
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs:46]: I don't think I've used a double await before, would love to learn more about this pattern and why we do this

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 15]
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
             return !_disposing && !_disposed && _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
         }
 
         public async Task StartWorkerProcessAsync(CancellationToken cancellationToken)
         {
-            _startSubscription = _inboundWorkerEvents.Where(msg => msg.MessageType == MsgType.StartStream)
-                .Timeout(_workerConfig.CountOptions.ProcessStartupTimeout)
-                .Take(1)
-                .Subscribe(SendWorkerInitRequest, HandleWorkerStartStreamError);
+            OnNext(MsgType.StartStream, _workerConfig.CountOptions.ProcessStartupTimeout, 1, SendWorkerInitRequest, HandleWorkerStartStreamError);
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Is this `OnNext()` method here to line up with the Rx subscriptions we had set up earlier? A more lightweight subscription?

**Review by brettsam — COMMENTED** [review]
I only have very general questions here. I think we should get this run on a private stamp to make sure we have no cold start regressions. 

I also want to take one more pass through to think through things like restarts and failures -- this is tricky!

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 15 above)
Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: I'm wondering if we can rename this to some thing more descriptive so its easier to follow, or the method that's using it

**Comment by liliankasem** [comment]
@mgravell anything we can do to help you get this PR merged?

File: src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs (hunk 14 above)
Comment (NickCraver) [reply src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs:46]: The reason there's a double `await` here is the `Task<Task<T>>` nature, so we're awaiting the async yield of the task itself, and then (outer) awaiting _that _ task.

**Comment by fabiocav** [comment]
@mgravell indeed, this is high priority on our side, so whatever we can do to help, please let us know.

**Comment by NickCraver** [comment]
I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.

**Comment by liliankasem** [comment]
> I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.

I think it's a mix of both, would be great to get PR feedback addressed but not a huge blocker. @safihamid and @pragnagopa can you help with cold start testing?

**Comment by safihamid** [comment]
> > I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.
> 
> I think it's a mix of both, would be great to get PR feedback addressed but not a huge blocker. @safihamid and @pragnagopa can you help with cold start testing?

Ok I will get cold start comparison numbers for this by end of next week. I hope most PR feedbacks are addressed by then.

File: src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs (hunk 5 above)
Comment (mgravell) [reply src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs]: sorry, must have been leftover from an earlier iteration; removing

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 6 above)
Comment (mgravell) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: no real benefit either way; happy to change, but... we know they're the same value

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 8 above)
Comment (mgravell) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: switch expressions require ... an expression - i.e. a single computed value for each outcome, that must all evaluate to some single type (perhaps after coercion). In this case, most of the things we actually do here are `void` returns. Forcing it to try to use a switch expression would be unnatural. If you mean a `when` clause, I could perhaps get behind that; as below:

``` c#
                switch (msg.MessageType)
                {
                    case MsgType.RpcLog when msg.Message.RpcLog.LogCategory == RpcLogCategory.System:
                        SystemLog(msg);
                        break;
                    case MsgType.RpcLog:
                        Log(msg);
                        break;
                    case MsgType.WorkerStatusResponse:
                        ReceiveWorkerStatusResponse(msg.Message.RequestId, msg.Message.WorkerStatusResponse);
                        break;
                    case MsgType.InvocationResponse:
                        _ = InvokeResponse(msg.Message.InvocationResponse);
                        break;
                    default:
                        OnNext(msg);
                        break;
                }
```

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 3 above)
Comment (mgravell) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: fixing

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 4 above)
Comment (mgravell) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:242]: They're both just "run me a thing" abstractions; QUWI works fine, and has fewer overheads (marginally) than Task.Run; either should work

**Comment by mgravell** [comment]
Merge is in progress; fighting `SendInvocationRequest_ValidateTraceContext` and `SendInvocationRequest_ValidateTraceContext_SessionId` currently

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 8 above)
Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Yup, that looks great - thanks!

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs (hunk 1 above)
Comment (safihamid) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:51]: ok I don't see any changes in cold start in my private stamp with this. 

this method (_eventManager.AddGrpcChannels(workerId))gets called during placeholder mode.

**Comment by safihamid** [comment]
No noticeable cold start changes with this PR in my private stamp as long as we are using the right language worker during placeholder mode. i.e. use Node placeholder for Node app, ...

**Comment by paulbatum** [comment]
I ran a few throughput tests in production using an App Service Plan running custom site extensions with and without the changes in this PR. 

TL;DR - while the gains for single core (P1V2) are marginal, the gains for multicore setups (P3V2 is 4-core) are significant, can be as high as 50% 🚀 🥇  

| Configuration | Throughput (dev) | Throughput (remove-rx) |
|-|-|-|
| P1V2, 100 connections | 401 RPS | 416 RPS |
| P3V2, 400 connections | 1042 RPS | 1581 RPS |

These tests were run with the following configuration:
- 30 second test runs
- workload is NET isolated anonymous HTTP triggered function that just returns a 200 response
- host.json is modified to reduce the amount of logging
- load is generated using bombadier e.g. 
`d:\tools\bombardier-windows-amd64.exe --insecure -c 400 -r 2400 -d 30s -m POST --format=pt -p i,r -l https://hostname.azurewebsites.net/api/function1`

dev branch latency distribution - 
```
Bombarding https://pbatum-rx-test-isolated.azurewebsites.net:443/api/function1 for 30s using 400 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      1042.60     200.06    1799.95
  Latency      380.22ms    36.21ms      1.28s
  Latency Distribution
     50%   376.40ms
     75%   388.20ms
     90%   408.00ms
     95%   429.02ms
     99%   506.01ms
  HTTP codes:
    1xx - 0, 2xx - 31666, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   585.32KB/s
```

no-rx latency distribution - 
```
Bombarding https://pbatum-rx-test-isolated.azurewebsites.net:443/api/function1 for 30s using 400 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      1581.55     940.41    8996.40
  Latency      248.55ms    53.08ms      0.85s
  Latency Distribution
     50%   236.14ms
     75%   277.61ms
     90%   325.00ms
     95%   365.00ms
     99%   474.09ms
  HTTP codes:
    1xx - 0, 2xx - 47739, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   845.26KB/s
```

**Comment by NickCraver** [comment]
Most failures here are due to more merges since the PR - 1 left I couldn't get through tonight:
```
[xUnit.net 00:03:48.32]     Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded [FAIL]
  Failed Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded [2 m]
  Error Message:
   System.ApplicationException : Condition not reached within timeout.
  Stack Trace:
     at Microsoft.Azure.WebJobs.Script.Tests.TestHelpers.Await(Func`1 condition, Int32 timeout, Int32 pollingInterval, Boolean throwWhenDebugging, Func`1 userMessageCallback) in /_/test/WebJobs.Script.Tests.Shared/TestHelpers.cs:line 115
   at Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded() in /_/test/WebJobs.Script.Tests.Integration/Rpc/WorkerConcurrencyManagerEndToEndTests.cs:line 28
--- End of stack trace from previous location ---
fail: Host.General[515]
      A host error has occurred during startup operation 'fe3a9a1d-e32b-4243-9d9f-2e2d3bcbbfe3'.
      System.ArgumentNullException: Value cannot be null. (Parameter 'provider')
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
         at Microsoft.Azure.WebJobs.Script.ScriptHostBuilderExtensions.<>c__DisplayClass6_0.<AddScriptHost>b__2(HostBuilderContext context, IConfigurationBuilder configBuilder) in /_/src/WebJobs.Script/ScriptHostBuilderExtensions.cs:line 114
         at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
         at Microsoft.Extensions.Hosting.HostBuilder.Build()
         at Microsoft.Azure.WebJobs.Script.WebHost.DefaultScriptHostBuilder.BuildHost(Boolean skipHostStartup, Boolean skipHostConfigurationParsing) in /_/src/WebJobs.Script.WebHost/DefaultScriptHostBuilder.cs:line 59
         at Microsoft.Azure.WebJobs.Script.WebHost.WebJobsScriptHostService.BuildHost(Boolean skipHostStartup, Boolean skipHostJsonConfiguration) in /_/src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:line 581
         at Microsoft.Azure.WebJobs.Script.WebHost.WebJobsScriptHostService.UnsynchronizedStartHostAsync(ScriptHostStartupOperation activeOperation, Int32 attemptCount, JobHostStartupMode startupMode) in /_/src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:line 276
```

I'm not setup for end-to-end on laptop so can't debug readily but something is amiss there with services - will poke more tomorrow.

**Review by liliankasem — APPROVED** [review]

**Review by paulbatum — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 16]
```diff
@@ -140,17 +134,133 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
+                        _ = InvokeResponse(msg.Message.InvocationResponse);
+                        break;
+                    default:
+                        ProcessRegisteredGrpcCallbacks(msg);
+                        break;
+                }
+            }
+            catch
+            {
+                // TODO: log?
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Yeah, let's add a log here so we can trace this if we get CRIs

**Review by liliankasem — COMMENTED** [review]
just noticed there's a TODO here, should probably resolve before merge! (+ any other TODOs I may have missed). If not addressed, you should create an issue for any items we're not handling in this PR

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 16 above)
Comment (NickCraver) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: @liliankasem Can you please add the logging intended here? I'm not sure what the expectations are, probably quicker to push change to branch directly. It'll happen on _every_ gRPC event so not sure how we best handle potentially large volume things.

Comment (fabiocav) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Let's not block. @liliankasem maybe we can log an issue to track and have a follow up PR with anything that might be needed.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 17]
```diff
@@ -859,9 +958,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
...
+            }
+            var evt = new OutboundGrpcEvent(_workerId, msg);
+            if (!_outbound.TryWrite(evt))
+            {
+                var pending = _outbound.WriteAsync(evt);
+                if (pending.IsCompleted)
+                {
+                    try
+                    {
+                        pending.GetAwaiter().GetResult(); // ensure observed, for IVTS reasons
```
Comment (fabiocav) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: nit: would it be helpful to update the comment to something more descriptive (e.g., `// ensure observed to ensure IValueTaskSource completed/result is consumed`)?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 18]
```diff
@@ -102,31 +105,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
...
-
-            _eventSubscriptions.Add(_inboundWorkerEvents
-                .Where(msg => msg.IsMessageOfType(MsgType.RpcLog) && msg.IsLogOfCategory(RpcLogCategory.System))
-                .Subscribe(SystemLog));
+            if (_eventManager.TryGetGrpcChannels(workerId, out var inbound, out var outbound))
+            {
+                _outbound = outbound.Writer;
+                _inbound = inbound.Reader;
+                // note: we don't start the read loop until StartWorkerProcessAsync is called
+            }
```
Comment (fabiocav) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:116]: The factory should create the channels prior to creating the `GrpcWorkerChannel`, so we should always have these channels setup here. Are expecting this to ever return `false`? Should we throw that is ever the case?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 19]
```diff
@@ -859,9 +958,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
             _eventManager.Publish(new WorkerErrorEvent(_runtime, Id, exc));
         }
 
+        private ValueTask SendStreamingMessageAsync(StreamingMessage msg)
+        {
+            if (_outbound is null)
+            {
+                return default;
```
Comment (fabiocav) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: For this scenario, particularly if the message was an invocation request, shouldn't we fail here? In the case of an invocation, will will the behavior be if we simply return? I'm assuming the invocation would remain in-flight until the function timeout (which could be quite large) occurs.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 16 above)
Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Done! #8604

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 17 above)
Comment (NickCraver) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Updates!

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 19 above)
Comment (NickCraver) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: With a change above, we shouldn't be able to get in this state at all - going to remove these extra checks since it's no longer nullable and checked in the constructor.

**Comment by NickCraver** [comment]
Gentle reminder we should try and get this in before more conflicts pile up against the old script manager approach.

**Review by brettsam — APPROVED** [review]

**Comment by brettsam** [comment]
Let's get it in -- no objections from me. Although I do see conflicts again...

**Comment by NickCraver** [comment]
@brettsam merged once more from dev and addressed one test race (existing issue but hey, improvements!)

**Review by brettsam — APPROVED** [review]

**Review by fabiocav — APPROVED** [review]

**Comment by fabiocav** [comment]
Thanks, @NickCraver.

Good to have this merged once conflicts are resolved.

**Comment by NickCraver** [comment]
@fabiocav I can't see merge conflicts local or PR UX, any idea what it's conflicting with?

**Comment by fabiocav** [comment]
@NickCraver odd, I don't see the usual list, just the conflict warning in GH.

If you can merge locally, go for it!

**Comment by NickCraver** [comment]
I beat the GitHub UI!
